src/hs/Vehicle/Core/Normalise.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DataKinds #-}
+
+module Vehicle.Core.Normalise
+  ( Norm
+  , NormError (..)
+  , MonadNorm
+  , runNorm
+  ) where
+
+import Vehicle.Core.Type ( Sort (..), Expr (..), Sort, EArg (..) )
+import Vehicle.Core.Instance.Recursive (ExprF (..))
+import Control.Monad.Except (MonadError, Except, runExcept)
+import Data.Map (Map, lookup, insert, notMember)
+import Data.Functor.Foldable (fold)
+import Vehicle.Core.Abs (ExprName, SortedBuiltin)
+import Control.Monad.Error.Class (throwError)
+
+-- * Normalisation monad
+
+type DeBruijn = Int
+
+newtype SortedDeBruijn (sort :: Sort) 
+  = SortedDeBruijn DeBruijn
+
+type NormExpr ann = Expr SortedDeBruijn SortedBuiltin ann
+
+-- |Errors thrown during normalisation
+data NormError = MissingDefFunType ExprName
+    | UnboundVariableError DeBruijn
+    | DuplicateVariableError DeBruijn
+    | MalformedLambdaError
+
+-- |Constraint for the monad stack used by the normaliser.
+type MonadNorm m = MonadError NormError m
+
+-- | Run a function in 'MonadNorm'.
+runNorm :: Except NormError a -> Either NormError a
+runNorm = runExcept
+
+-- |Class for the various normalisation functions.
+-- Invariant is that everything in the context is fully normalised
+class Norm vf where
+  norm :: MonadNorm m => Map DeBruijn vf -> vf -> m vf
+
+instance Norm (NormExpr ann) where
+  norm ctxt = fold $ \case
+    EVarF ann (SortedDeBruijn i) -> normVar ctxt ann i
+    ELetF _ arg exp1 exp2 -> normLet ctxt (index arg) exp1 exp2
+    EAppF _ exp1 exp2 -> normApp ctxt exp1 exp2
+
+    ELitSeqF ann exprs -> ELitSeq ann <$> sequence exprs
+    EAnnF ann expr typ -> EAnn ann <$> expr <*> pure typ
```

Comment:
Is there a function that does `\v -> _<*> pure v` automatically? Couldn't find one...